toolchain/check/pointer_dereference.cpp

```diff
@@ -20,13 +19,6 @@ auto PerformPointerDereference(Context& context,
   if (auto pointer_type =
           context.types().TryGetAs<SemIR::PointerType>(type_id)) {
     result_type_id = pointer_type->pointee_id;
-  } else if (type_id != SemIR::TypeId::Error) {
-    CARBON_DIAGNOSTIC(DerefOfNonPointer, Error,
-                      "Cannot dereference operand of non-pointer type `{0}`.",
-                      SemIR::TypeId);
-    auto builder =
-        context.emitter().Build(TokenOnly(node_id), DerefOfNonPointer, type_id);
-    builder.Emit();
   }
```

Comment:
For the case of `not_pointer->member`, we should produce a diagnostic of some kind.

Perhaps you could make `PerformPointerDereference` take an `llvm::function_ref<auto (SemIR::TypeId not_pointer_type_id) -> void> diagnose_not_pointer` parameter and call it here if the type isn't a pointer type, and pass a lambda that emits a diagnostic from each of the two callers. That way we can avoid the duplicated check in `HandlePrefixOperatorStar`, and emit a custom diagnostic from `HandlePointerMemberAccessExpr` -- perhaps something like "Cannot apply `->` operator to non-pointer type `{0}`."

We should also make sure that case is tested.

(It'd be cool for the `->` case to recover by treating the `->` operator as a `.` instead, especially if we find that the left-hand operand's type has a member with the specified name, but I think that's too much complexity to add in this patch. So let's not do that in this PR.)